test/MultiThreadedCaches.jl

```diff
@@ -0,0 +1,76 @@
+using MultiThreadedCaches
+
+using Test
+
+@info "Testing with:" Threads.nthreads()
+
+@testset "basics" begin
+    cache = MultiThreadedCache{Int,Int}()
+    init_cache!(cache)
+
+    @test length(cache.thread_caches) == Threads.nthreads()
+
+    get!(cache, 1) do
+        return 10
+    end
+
+    # The second get!() shouldn't have any effect, and the result should still be 10.
+    @test get!(cache, 1) do
+        return 100
+    end == 10
+
+    # Even on different threads, the cache should still have a value for key `1`.
+    @sync for i in 1:100
+        Threads.@spawn begin
+            @test get!(cache, 1) do
+                return 100
+            end == 10
+        end
+    end
+
+    # Internals test:
+    @test length(cache.base_cache_futures) == 0
+end
+
+# Helper function for stress-test: returns true if all elements in iterable `v` are equal.
+function all_equal(v)
+    length(v) < 2 && return true
+    e1 = first(v)
+    i = 2
+    @inbounds for i=2:length(v)
+        v[i] == e1 || return false
+    end
+    return true
+end
+
+@testset "stress-test" begin
+    cache = MultiThreadedCache{Int,Int}()
+    init_cache!(cache)
+
+    len = 100
+    outputs = [Channel{Int}(Inf) for _ in 1:len]
+
+    @sync for i in 1:100_000
+        Threads.@spawn begin
```

Comment:
You have a correctness test here, but there's no perf test. I suggest using `@elapsed` and running this stress test twice; once with just 1 thread (no `@spawn`) and once with multiple threads. The comparison will show scalability.

Reply:
yeah, good idea. Should that go in `test`? Or should we have a separate benchmarks directory?

Reply:
Unless you plan on expanding the feature set dramatically, there's probably not much scope for a lot more tests or benchmarks? I'd put it here for now and split it out later if needed, but it's your call. :)

Reply:
Actually, Kiran, there's no scaling benefit here to running in multiple threads...

The scaling benefit comes from _caching_, not from parallelism here. The thread-safety requirement comes from the fact that you want to use those cached values even across multiple threads. The goal of this package is to provide that thread safety without adding too much contention that would _reduce_ performance as you scale. 

If everything is working well here, this test should have close to the same performance as running single-threaded, but certainly wouldn't be any faster. Does that make sense?

Reply:
Yes, what you want to verify is that the performance is good even in the presence of contention.

Reply:
gotcha. It seems to be, and I'll just have the benchmark print out the results. 👍 Thanks

Reply:
Here's a PR for the benchmark numbers: https://github.com/NHDaly/MultiThreadedCaches.jl/pull/4